coil-gif/src/main/java/coil/decode/ImageDecoderDecoder.kt

```diff
@@ -43,15 +44,56 @@ class ImageDecoderDecoder @JvmOverloads constructor(
     private val enforceMinimumFrameDelay: Boolean = true
 ) : Decoder {
 
-    override suspend fun decode(): DecodeResult {
+    override suspend fun decode(): DecodeResult = decode2()
+
+    private fun ImageSource.toImageDecoderSource(): ImageDecoder.Source {
+        val file = fileOrNull()
+        if (file != null) {
+            return ImageDecoder.createSource(file)
+        }
+
+        val metadata = metadata
+        if (metadata is AssetMetadata) {
+            return ImageDecoder.createSource(options.context.assets, metadata.fileName)
+        }
+        if (metadata is ContentMetadata) {
+            return ImageDecoder.createSource(options.context.contentResolver, metadata.uri)
+        }
+        if (metadata is ResourceMetadata && metadata.packageName == options.context.packageName) {
+            return ImageDecoder.createSource(options.context.resources, metadata.resId)
+        }
+        return when {
+            SDK_INT >= 31 -> ImageDecoder.createSource(source().use { it.readByteArray() })
+            SDK_INT == 30 -> ImageDecoder.createSource(
+                ByteBuffer.wrap(source().use { it.readByteArray() })
+            )
+            // https://issuetracker.google.com/issues/139371066
+            else -> ImageDecoder.createSource(file())
+        }
+    }
+
+    private fun getImageDecoderSource(imageSource: ImageSource): ImageSource {
```

Comment:
This change makes it so `imageSource` is never closed.

Reply:
This method doesn't deal with closing the `resource`. The idea is to only composite the `ImageSource` and the code that uses the resource becomes the responsible for closing it. The `ImageSource` will be closed later in `toImageDecoderSource`.